### PR TITLE
Add support for querying noDict MV columns for offline (all data types) and realtime (fixed width) segments

### PR DESCRIPTION
**Pull request by somandal**
This PR adds support for querying multi-value columns in RAW format (without dictionary). Support to create segments with multi-value columns with dictionary disabled already existed. On trying to create a multi-value RAW column and querying it, we ran into this issue: https://github.com/apache/pinot/issues/8875

This PR adds MV RAW support for the following types of columns:

- Offline segments with MV columns of types INT, LONG, FLOAT, DOUBLE, and STRING
- Mutable (realtime) segments with MV columns of INT, LONG, FLOAT, and DOUBLE

TODO - Support for variable length MV RAW columns for Mutable segments will be added in the future as it requires a mutable variable length writer.

Test cases for various types of queries have been added to ensure that MV RAW columns can be supported.

cc @siddharthteotia

### CODE REVIEW

**Comment by siddharthteotia** [comment]
hi @Jackie-Jiang  / @richardstartin - I had iterated upon the review internally on the branch with @somandal  before the PR was sent out. It will be good if one of you can also help review. Thanks

File: pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/reader/ForwardIndexReader.java [hunk 1]
```diff
@@ -47,7 +48,7 @@
    * Returns the data type of the values in the forward index. Returns {@link DataType#INT} for dictionary-encoded
    * forward index.
    */
-  DataType getValueType();
+  DataType getStoredType();
```
Comment (Jackie-Jiang) [inline pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/reader/ForwardIndexReader.java:51]: Can we put this change as a separate PR?

File: pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/reader/ForwardIndexReader.java [hunk 2]
```diff
@@ -410,9 +411,318 @@ default byte[] getBytes(int docId, T context) {
...
+   * @param maxNumValuesPerMVEntry maximum number of values per MV entry
+   * @param values Values to fill
+   * @param context Reader context
+   */
+  default void readValuesMV(int[] docIds, int length, int maxNumValuesPerMVEntry, int[][] values, T context) {
+    switch (getStoredType()) {
+      case INT:
+        int[] intValueBuffer = new int[maxNumValuesPerMVEntry];
+        for (int i = 0; i < length; i++) {
+          int numValues = getIntMV(docIds[i], intValueBuffer, context);
```
Comment (Jackie-Jiang) [inline pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/reader/ForwardIndexReader.java:429]: For MV read, let's add APIs to directly read the values without passing in a buffer, e.g. `int[] getIntMV(int docId, T context)`. This API will be very useful to prevent unnecessary copying of the array, or when the `maxNumValuesPerMVEntry` is not available

File: pinot-core/src/main/java/org/apache/pinot/core/common/DataFetcher.java [hunk 3]
```diff
@@ -713,8 +730,51 @@ void readStringValuesMV(TransformEvaluator evaluator, int[] docIds, int length,
...
+      if (_dictionary != null) {
+        for (int i = 0; i < length; i++) {
+          numValuesBuffer[i] = _reader.getDictIdMV(docIds[i], _reusableMVDictIds, getReaderContext());
+        }
+      } else {
+        switch (_reader.getStoredType()) {
+          case INT:
+            int[] intValueBuffer = new int[_maxNumValuesPerMVEntry];
+            for (int i = 0; i < length; i++) {
+              numValuesBuffer[i] = _reader.getIntMV(docIds[i], intValueBuffer, getReaderContext());
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/common/DataFetcher.java:742]: This is adding lots of overhead because we don't really need to read the values. Let's add an API `int getNumValuesMV(int docId, T context)` to the `ForwardIndexReader` which simply returns the values in the MV entry without reading any content

**Review by Jackie-Jiang — COMMENTED** [review]
This is great, thanks for adding all the tests!

This PR is a little bit too large, and contains changes to lots of modules. I'd suggest breaking them into smaller ones so that each PR only focus on one thing. That can make both review and future tracking easier.
We can break the PR into the following smaller scopes:
- Change `getValueType` to `getStoredType`
- New APIs to ForwardIndexReader and the implementations
- Filter support / group by support etc.

**Comment by siddharthteotia** [comment]
> This is great, thanks for adding all the tests!
> 
> This PR is a little bit too large, and contains changes to lots of modules. I'd suggest breaking them into smaller ones so that each PR only focus on one thing. That can make both review and future tracking easier. We can break the PR into the following smaller scopes:
> 
> * Change `getValueType` to `getStoredType`
> * New APIs to ForwardIndexReader and the implementations
> * Filter support / group by support etc.

Sounds good. When we were iterating upon it internally, it was more of what all is needed to get this done correctly throughout the code base with all testing so there was some back n forth. Now that we agree with the overall work, I am ok with the above suggestion to breakdown.  cc @somandal

File: pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/reader/ForwardIndexReader.java (hunk 2 above)
Comment (somandal) [reply pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/reader/ForwardIndexReader.java:429]: Just to clarify, for the types where we need to typecast or return the values[] of a different stored type, I will still need to copy the data into the values buffer. Is this copying still okay?

**Comment by somandal** [comment]
> This PR adds support for querying multi-value columns in RAW format (without dictionary). Support to create segments with multi-value columns with dictionary disabled already existed. On trying to create a multi-value RAW column and querying it, we ran into this issue: #8875
> 
> This PR adds MV RAW support for the following types of columns:
> 
> * Offline segments with MV columns of types INT, LONG, FLOAT, DOUBLE, and STRING
> * Mutable (realtime) segments with MV columns of INT, LONG, FLOAT, and DOUBLE
> 
> TODO - Support for variable length MV RAW columns for Mutable segments will be added in the future as it requires a mutable variable length writer.
> 
> Test cases for various types of queries have been added to ensure that MV RAW columns can be supported.
> 
> cc @siddharthteotia

Thanks for taking a look and for the suggestion to break this down. I've got the first part out for review in this PR: https://github.com/apache/pinot/pull/8953


> This is great, thanks for adding all the tests!
> 
> This PR is a little bit too large, and contains changes to lots of modules. I'd suggest breaking them into smaller ones so that each PR only focus on one thing. That can make both review and future tracking easier. We can break the PR into the following smaller scopes:
> 
> * Change `getValueType` to `getStoredType`
> * New APIs to ForwardIndexReader and the implementations
> * Filter support / group by support etc.

Thanks for the suggestion @Jackie-Jiang! I've opened a new PR for the second one on your list "New APIs to ForwardIndexReader and the implementations" in this PR: https://github.com/apache/pinot/pull/8953. Can you take a look when you get a chance? (I've also tried to address your comments on that PR directly)

I'll be working on the `getStoredType` after the other two.

File: pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/reader/ForwardIndexReader.java (hunk 2 above)
Comment (Jackie-Jiang) [reply pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/reader/ForwardIndexReader.java:429]: Yes. If the buffer can be reused, we should still create a buffer to reduce the garbage, and that's why we want to have 2 versions of the API, one with buffer and one without buffer

**Comment by siddharthteotia** [comment]
Part 1 https://github.com/apache/pinot/pull/8953 split from this PR has been merged. Rest of the work to follow-up

**Comment by siddharthteotia** [comment]
Part 2 split from this PR https://github.com/apache/pinot/pull/8993
